File: python/a30.py
# A30 Combination
# submission: https://atcoder.jp/contests/tessoku-book/submissions/38582866

MODULO = 10 ** 9 + 7

def factorial(n: int) -> int:
    ret = 1
    for i in range(1, n + 1):
        ret *= i
        ret %= MODULO
    return ret

def power(x: int, y: int) -> int:
    ret = 1
    tmp = x
    for i in range(30):
        if y & (1 << i):
            ret *= tmp
            ret %= MODULO
        tmp **= 2
        tmp %= MODULO
    return ret

def inverse(b: int) -> int:
    b %= MODULO
    return power(b, MODULO - 2)

n, r = map(int, input().split())

print((factorial(n) * inverse(factorial(r) * factorial(n - r))) % MODULO)

File: rust/tessoku-book/src/bin/a30.rs
// A30 Combination
// submission: https://atcoder.jp/contests/tessoku-book/submissions/38586409

use proconio::input;

const MODULO: u64 = 1_000_000_007;

fn factorial(n: u64) -> u64 {
    let mut ret = 1;
    for i in 1..=n {
        ret *= i;
        ret %= MODULO;
    }
    ret
}

fn power(x: u64, y: u64) -> u64 {
    let mut ret = 1;
    let mut tmp = x;
    for i in 0..30 {
        if y & (1 << i) > 0 {
            ret *= tmp;
            ret %= MODULO;
        }
        tmp = tmp * tmp;
        tmp %= MODULO;
    }
    return ret;
}

fn inverse(b: u64) -> u64 {
    let bb = b % MODULO;
    power(bb, MODULO - 2)
}

fn main() {
    input!{
        n: u64,
        r: u64,
    }
    println!("{}", (factorial(n) * inverse(factorial(r) * factorial(n - r))) % MODULO);
}
